### PR TITLE
[WIP] validate resource version specified in the request URI

### PR DESCRIPTION
**Pull request by tkashem**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
Validate resource version specified in the request URI. If a user specifies an invalid resourceVersion in the request URI, instead of receiving a HTTP 500 the caller will receive an HTTP status code 400. 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
If a user specifies an invalid resourceVersion in the request URI, instead of receiving a HTTP 500 the caller will receive an HTTP status code 400. 
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by leilajal** [comment]
cc @lavalamp

**Comment by leilajal** [comment]
cc @yliaog @deads2k

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/resource_version.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	"k8s.io/apiserver/pkg/endpoints/request"
+)
+
+// WithResourceVersionValidation ensures that the "resource version" query prameter
+// if specified is valid.
+// If an invalid value of resourceVersion is specified in the request URI, the
+// request is rejected with a 400.
+func WithResourceVersionValidation(handler http.Handler, negotiatedSerializer runtime.NegotiatedSerializer) http.Handler {
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/resource_version.go:35]: The field is a string, all strings are valid from a schema perspective. Not all requests have a resource version parameter, so we should not use a filter to accomplish this anyway.

**Comment by lavalamp** [comment]
I buy that 400 is the right response, but it should be solved by taking greater care plumbing the existing error.

**Comment by tkashem** [comment]
holding it until I get a buy in to move forward with an agreeable solution  cc @sttts

/hold

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/resource_version.go (hunk 1 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/resource_version.go:35]: yeah, usually the validation happens in [the handlers ](https://github.com/kubernetes/kubernetes/blob/5d61e18a4343773105e39c36b944b4f223e70eaf/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/get.go#L203)or even directly in [the registry](https://github.com/kubernetes/kubernetes/blob/8b98305858b107369f2c9b9fd8ef1c5b0da078c0/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L819). We should have something similar for checking the `resourceVersion`.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/resource_version.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+func hasValidResourceVersion(r *http.Request) bool {
+	rv := r.URL.Query().Get("resourceVersion")
+	if rv == "" {
+		// resourceVersion can be empty.
+		return true
+	}
+
+	if _, err := strconv.ParseInt(rv, 0, 64); err != nil {
```
Comment (yliaog) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/resource_version.go:63]: it is not an integer, it just happens to be an int for now, but any string should just work.

**Comment by p0lyn0mial** [comment]
@tkashem Have you checked what kind of error we get from etcd when an incorrect `rv` is provided?
Maybe we could translate it to something that would help us to return `400 `.

**Comment by lavalamp** [comment]
Happy to see a version of this that interprets the error in the right place, but we can't fix it this way.

**Pull request closed without merging**